File: core/gesture_engine/gesture_engine.cpp
/**
 * gesture_engine.cpp — Phase 2
 *
 * Full state machine implementation for gesture recognition.
 * See §4.6 of the execution plan for the complete state model.
 */
#include "gesture_engine.h"
#include <cmath>
#include <algorithm>
#include <iostream>

namespace magicmouse {

GestureEngine::GestureEngine(GestureCallback cb, GestureParams params)
    : callback_(std::move(cb)), params_(params) {}

void GestureEngine::onTouchFrame(const TouchPoint* points, int count, int64_t ts_ms) {
    if (count == 0) return;

    float x = points[0].x;
    float y = points[0].y;

    switch (state_) {
    case State::IDLE:
    case State::TOUCH_START:
        start_x_ = x; start_y_ = y;
        last_x_ = x;  last_y_ = y;
        velocity_x_ = 0; velocity_y_ = 0;
        finger_count_ = count;
        touch_start_ms_ = ts_ms;
        transition(State::TRACKING);
        break;

    case State::TRACKING: {
        float dx = x - last_x_;
        float dy = y - last_y_;
        velocity_x_ = dx;
        velocity_y_ = dy;

        float total_dx = std::abs(x - start_x_);
        float total_dy = std::abs(y - start_y_);

        // Enter SCROLL_LOCK if vertical/horizontal movement dominates
        if (total_dy >= params_.scroll_lock_threshold_px && total_dy > total_dx) {
            transition(State::SCROLL_LOCK);
            emitScroll(0.0f, dy);
        } else if (total_dx >= params_.scroll_lock_threshold_px && total_dx > total_dy) {
            transition(State::SCROLL_LOCK);
            emitScroll(dx, 0.0f);
        }
        // Gesture candidate if swipe-like movement
        else if (total_dx >= params_.min_swipe_distance_px || total_dy >= params_.min_swipe_distance_px) {
            int64_t elapsed = ts_ms - touch_start_ms_;
            if (elapsed < params_.max_swipe_time_ms) {
                transition(State::GESTURE_CANDIDATE);
            }
        }
        last_x_ = x; last_y_ = y;
        break;
    }

    case State::SCROLL_LOCK: {
        float dx = x - last_x_;
        float dy = y - last_y_;
        emitScroll(dx, dy);
        last_x_ = x; last_y_ = y;
        break;
    }

    case State::GESTURE_CANDIDATE:
        classify(ts_ms);
        break;

    default:
        break;
    }
}

void GestureEngine::onFingerLift(int64_t ts_ms) {
    if (state_ == State::GESTURE_CANDIDATE || state_ == State::GESTURE_AMBIGUOUS) {
        classify(ts_ms);
    } else if (state_ == State::TRACKING) {
        // Tap detection
        float total_move = std::hypot(last_x_ - start_x_, last_y_ - start_y_);
        if (total_move <= params_.tap_max_movement_px) {
            int64_t since_last_lift = ts_ms - last_lift_ms_;
            if (since_last_lift <= params_.double_tap_interval_ms) {
                // Double tap
                GestureEvent ev;
                ev.type = GestureType::DOUBLE_TAP;
                ev.fingers = finger_count_;
                if (callback_) callback_(ev);
                tap_count_ = 0;
            } else {
                // Single/multi-finger tap based on finger count
                GestureEvent ev;
                ev.fingers = finger_count_;
                if (finger_count_ == 1)      ev.type = GestureType::SINGLE_TAP;
                else if (finger_count_ == 2) ev.type = GestureType::TWO_FINGER_TAP;
                else if (finger_count_ >= 3) ev.type = GestureType::THREE_FINGER_TAP;
                if (callback_) callback_(ev);
            }
        }
    }

    last_lift_ms_ = ts_ms;
    transition(State::IDLE);
}

void GestureEngine::transition(State next) {
    state_ = next;
}

void GestureEngine::classify(int64_t ts_ms) {
    float dx = last_x_ - start_x_;
    float dy = last_y_ - start_y_;
    float adx = std::abs(dx);
    float ady = std::abs(dy);

    // Check gesture_lock_ms: suppress gestures if we just came off a scroll
    int64_t since_lift = ts_ms - last_lift_ms_;
    if (since_lift < params_.gesture_lock_ms) {
        transition(State::IDLE);
        return;
    }

    // Velocity dominance check
    bool h_dominant = (adx > ady) &&
        (velocity_x_ != 0) &&
        (std::abs(velocity_x_) / (std::abs(velocity_y_) + 0.001f) >= params_.gesture_velocity_ratio);
    bool v_dominant = (ady > adx) &&
        (velocity_y_ != 0) &&
        (std::abs(velocity_y_) / (std::abs(velocity_x_) + 0.001f) >= params_.gesture_velocity_ratio);

    GestureEvent ev;
    if (h_dominant && adx >= params_.min_swipe_distance_px) {
        ev.type = (dx > 0) ? GestureType::SWIPE_RIGHT : GestureType::SWIPE_LEFT;
    } else if (v_dominant && ady >= params_.min_swipe_distance_px) {
        ev.type = (dy > 0) ? GestureType::SWIPE_DOWN : GestureType::SWIPE_UP;
    } else {
        ev.type = GestureType::NONE;
    }

    ev.fingers = finger_count_;
    if (ev.type != GestureType::NONE && callback_) callback_(ev);
    transition(State::IDLE);
}

void GestureEngine::emitScroll(float dx, float dy) {
    if (dy != 0.0f) {
        GestureEvent ev;
        ev.type = GestureType::SCROLL_V;
        ev.delta = dy;
        ev.fingers = finger_count_;
        if (callback_) callback_(ev);
    }
    if (dx != 0.0f) {
        GestureEvent ev;
        ev.type = GestureType::SCROLL_H;
        ev.delta = dx;
        ev.fingers = finger_count_;
        if (callback_) callback_(ev);
    }
}

} // namespace magicmouse

File: service_rust/src/gesture_engine.rs
#[derive(Clone, Copy, Debug)]
pub struct TouchPoint {
    pub id: i32,
    pub x: f32,
    pub y: f32,
    pub pressure: f32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GestureType {
    None,
    SwipeLeft,
    SwipeRight,
    SwipeUp,
    SwipeDown,
    SingleTap,
    TwoFingerTap,
    ThreeFingerTap,
    DoubleTap,
    ScrollV,
    ScrollH,
}

#[derive(Clone, Copy, Debug)]
pub struct GestureEvent {
    pub gtype: GestureType,
    pub delta: f32,
    pub fingers: i32,
}

#[derive(Clone, Copy, Debug)]
pub struct GestureParams {
    pub min_swipe_distance_px: i32,
    pub max_swipe_time_ms: i32,
    pub scroll_lock_threshold_px: i32,
    pub gesture_velocity_ratio: f32,
    pub double_tap_interval_ms: i32,
    pub tap_max_movement_px: i32,
    pub gesture_lock_ms: i32,
}

impl Default for GestureParams {
    fn default() -> Self {
        Self {
            min_swipe_distance_px: 8,
            max_swipe_time_ms: 400,
            scroll_lock_threshold_px: 3,
            gesture_velocity_ratio: 2.0,
            double_tap_interval_ms: 250,
            tap_max_movement_px: 3,
            gesture_lock_ms: 300,
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum State {
    Idle,
    TouchStart,
    Tracking,
    ScrollLock,
    GestureCandidate,
    GestureAmbiguous,
}

pub struct GestureEngine<F>
where
    F: Fn(GestureEvent),
{
    callback: F,
    params: GestureParams,
    state: State,

    start_x: f32,
    start_y: f32,
    last_x: f32,
    last_y: f32,
    velocity_x: f32,
    velocity_y: f32,
    touch_start_ms: i64,
    last_lift_ms: i64,
    finger_count: i32,
    tap_count: i32,
}

impl<F> GestureEngine<F>
where
    F: Fn(GestureEvent),
{
    pub fn new(cb: F, params: GestureParams) -> Self {
        Self {
            callback: cb,
            params,
            state: State::Idle,
            start_x: 0.0,
            start_y: 0.0,
            last_x: 0.0,
            last_y: 0.0,
            velocity_x: 0.0,
            velocity_y: 0.0,
            touch_start_ms: 0,
            last_lift_ms: -10000,
            finger_count: 0,
            tap_count: 0,
        }
    }

    pub fn set_params(&mut self, p: GestureParams) {
        self.params = p;
    }

    pub fn on_touch_frame(&mut self, points: &[TouchPoint], ts_ms: i64) {
        if points.is_empty() { return; }

        let x = points[0].x;
        let y = points[0].y;
        let count = points.len() as i32;

        match self.state {
            State::Idle | State::TouchStart => {
                self.start_x = x;
                self.start_y = y;
                self.last_x = x;
                self.last_y = y;
                self.velocity_x = 0.0;
                self.velocity_y = 0.0;
                self.finger_count = count;
                self.touch_start_ms = ts_ms;
                self.state = State::Tracking;
            }
            State::Tracking => {
                let dx = x - self.last_x;
                let dy = y - self.last_y;
                self.velocity_x = dx;
                self.velocity_y = dy;

                let total_dx = (x - self.start_x).abs();
                let total_dy = (y - self.start_y).abs();

                if total_dy >= self.params.scroll_lock_threshold_px as f32 && total_dy > total_dx {
                    self.state = State::ScrollLock;
                    self.emit_scroll(0.0, dy);
                } else if total_dx >= self.params.scroll_lock_threshold_px as f32 && total_dx > total_dy {
                    self.state = State::ScrollLock;
                    self.emit_scroll(dx, 0.0);
                } else if total_dx >= self.params.min_swipe_distance_px as f32 || total_dy >= self.params.min_swipe_distance_px as f32 {
                    let elapsed = ts_ms - self.touch_start_ms;
                    if elapsed < self.params.max_swipe_time_ms as i64 {
                        self.state = State::GestureCandidate;
                    }
                }

                self.last_x = x;
                self.last_y = y;
            }
            State::ScrollLock => {
                let dx = x - self.last_x;
                let dy = y - self.last_y;
                self.emit_scroll(dx, dy);
                self.last_x = x;
                self.last_y = y;
            }
            State::GestureCandidate => {
                self.classify(ts_ms);
            }
            _ => {}
        }
    }

    pub fn on_finger_lift(&mut self, ts_ms: i64) {
        if self.state == State::GestureCandidate || self.state == State::GestureAmbiguous {
            self.classify(ts_ms);
        } else if self.state == State::Tracking {
            let dx = self.last_x - self.start_x;
            let dy = self.last_y - self.start_y;
            let total_move = (dx * dx + dy * dy).sqrt();

            if total_move <= self.params.tap_max_movement_px as f32 {
                let since_last = ts_ms - self.last_lift_ms;
                if since_last <= self.params.double_tap_interval_ms as i64 {
                    (self.callback)(GestureEvent {
                        gtype: GestureType::DoubleTap,
                        delta: 0.0,
                        fingers: self.finger_count,
                    });
                    self.tap_count = 0;
                } else {
                    let gtype = match self.finger_count {
                        1 => GestureType::SingleTap,
                        2 => GestureType::TwoFingerTap,
                        _ => GestureType::ThreeFingerTap,
                    };
                    (self.callback)(GestureEvent {
                        gtype,
                        delta: 0.0,
                        fingers: self.finger_count,
                    });
                }
            }
        }

        self.last_lift_ms = ts_ms;
        self.state = State::Idle;
    }

    fn classify(&mut self, ts_ms: i64) {
        let dx = self.last_x - self.start_x;
        let dy = self.last_y - self.start_y;
        let adx = dx.abs();
        let ady = dy.abs();

        let since_lift = ts_ms - self.last_lift_ms;
        if since_lift < self.params.gesture_lock_ms as i64 {
            self.state = State::Idle;
            return;
        }

        let h_dominant = (adx > ady) && (self.velocity_x != 0.0) &&
            ((self.velocity_x.abs() / (self.velocity_y.abs() + 0.001)) >= self.params.gesture_velocity_ratio);
        
        let v_dominant = (ady > adx) && (self.velocity_y != 0.0) &&
            ((self.velocity_y.abs() / (self.velocity_x.abs() + 0.001)) >= self.params.gesture_velocity_ratio);

        let mut gtype = GestureType::None;

        if h_dominant && adx >= self.params.min_swipe_distance_px as f32 {
            gtype = if dx > 0.0 { GestureType::SwipeRight } else { GestureType::SwipeLeft };
        } else if v_dominant && ady >= self.params.min_swipe_distance_px as f32 {
            gtype = if dy > 0.0 { GestureType::SwipeDown } else { GestureType::SwipeUp };
        }

        if gtype != GestureType::None {
            (self.callback)(GestureEvent {
                gtype,
                delta: 0.0,
                fingers: self.finger_count,
            });
        }
        self.state = State::Idle;
    }

    fn emit_scroll(&self, dx: f32, dy: f32) {
        if dy != 0.0 {
            (self.callback)(GestureEvent {
                gtype: GestureType::ScrollV,
                delta: dy,
                fingers: self.finger_count,
            });
        }
        if dx != 0.0 {
            (self.callback)(GestureEvent {
                gtype: GestureType::ScrollH,
                delta: dx,
                fingers: self.finger_count,
            });
        }
    }
}
